### PR TITLE
Rails 6.1.4.7

### PR DESCRIPTION
**Pull request by Hamms**
In addition to the required dependency updates, also apply a few changes to the codebase that can't effectively be applied in advance:

- Generate a new `dashboard/config/storage.yml` with `rails app:update`. We aren't actually using any `ActiveStorage` functionality anywhere, but Rails 6.1 still expects this configuration file to be present
- Update `schema.rb`:
  - Update the null check syntax used by some virtual columns
  - Update `create_table` invocations to match new expected options format in Rails 6.1
- Fix violations of new Rubocop rule [Rails/CompactBlank](https://docs.rubocop.org/rubocop-rails/cops_rails.html#railscompactblank)
- Update our `no_transform_paths` customization to reflect changes to `ActionDispatch::FileHandler`
  - Add some new parameters to the `serve` method
  - Stub [the new `file_readable?` method](https://github.com/rails/rails/blob/66073335585f04f2ed0f5ef930eb3c8955d50a6a/actionpack/lib/action_dispatch/middleware/static.rb#L139-L145) in order to create a mock test route instead of [the removed `match?` method](https://github.com/rails/rails/blob/v6.0.4.4/actionpack/lib/action_dispatch/middleware/static.rb#L24-L47)
- Update some query count tests to reflect new (lower!) query counts after updating to ActiveRecord 6.1
- Update FakeDashboard's patching of `ActiveRecord::ConnectionAdapters::Mysql2Adapter` to reflect changes to its internal structure

## Links

- [Upgrading from Rails 6.0 to Rails 6.1](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#upgrading-from-rails-6-0-to-rails-6-1)
- https://github.com/code-dot-org/code-dot-org/pull/50693
- https://github.com/code-dot-org/code-dot-org/pull/50767
- https://github.com/code-dot-org/code-dot-org/pull/50768
- https://github.com/code-dot-org/code-dot-org/pull/50829
- https://github.com/code-dot-org/code-dot-org/pull/50910
- https://github.com/code-dot-org/code-dot-org/pull/50908

## Testing story

Largely relying on existing tests to ensure that this change doesn't disrupt any existing functionality. Also read through the major changes in this version (linked above) to confirm that none of them are immediately relevant for our usage.

## Follow-up work

After this has been deployed, we can remove [this transitional framework configuration](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/config/initializers/new_framework_defaults_6_1.rb) and start working on preparing our configuration for Rails 7.0

### CODE REVIEW

File: dashboard/config/storage.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+local:
+  service: Disk
+  root: <%= Rails.root.join("storage") %>
+
+# Use rails credentials:edit to set the AWS secrets (as aws:access_key_id|secret_access_key)
+# amazon:
+#   service: S3
+#   access_key_id: <%= Rails.application.credentials.dig(:aws, :access_key_id) %>
+#   secret_access_key: <%= Rails.application.credentials.dig(:aws, :secret_access_key) %>
+#   region: us-east-1
```
Comment (sureshc) [inline dashboard/config/storage.yml:14]: Can we grab the region dynamically? If the `CDO` object has been initialized at this point, then we could set it from `CDO.aws_region`

**Review by sureshc — APPROVED** [review]

File: dashboard/config/storage.yml (hunk 1 above)
Comment (Hamms) [reply dashboard/config/storage.yml:14]: We possibly could if we want to, but this is a commented-out section of an auto-generated configuration file which we don't actually use anywhere. It's just a default placeholder which Rails automatically provides for new projects as of v6.0 and which it expects to always be in place as of v6.1, it's not something we need to personalize until we decide to actually start using it.

File: dashboard/app/controllers/levels_controller.rb [hunk 2]
```diff
@@ -534,7 +534,7 @@ def embed_level
     # Reference links should be stored as an array.
     if params[:level][:reference_links].is_a? String
       params[:level][:reference_links] = params[:level][:reference_links].split("\r\n")
-      params[:level][:reference_links].delete_if(&:blank?)
+      params[:level][:reference_links].compact_blank!
```
Comment (davidsbailey) [inline dashboard/app/controllers/levels_controller.rb:537]: I am just curious, is this being forced by a newly enabled lint rule?

**Review by davidsbailey — APPROVED** [review]
LGTM! impressively small diff (if you ignore schema.rb). nice job getting lots of work done ahead in other PRs.

File: dashboard/app/controllers/levels_controller.rb (hunk 2 above)
Comment (Hamms) [reply dashboard/app/controllers/levels_controller.rb:537]: Yep: [Rails/CompactBlank](https://docs.rubocop.org/rubocop-rails/cops_rails.html#railscompactblank)

File: dashboard/config/storage.yml (hunk 1 above)
Comment (sureshc) [reply dashboard/config/storage.yml:14]: oh, my. I need to improve my reading comprehension. 😂
Yup. these placeholders are fine.

**Pull request merged**